File: backend/app/schemas/character_asset.py
"""CharacterAsset v1 validation shared by Package and world snapshot boundaries."""

import re
from urllib.parse import urlparse


SCHEMA_VERSION = "character-asset.v1"
_SHA256 = re.compile(r"^[a-fA-F0-9]{64}$")
_SAFE_RELATIVE_PREFIXES = ("assets/", "models/")
_SAFE_ABSOLUTE_PREFIXES = ("/assets/", "/models/", "/api/v0/assets/")


class CharacterAssetSchemaError(ValueError):
    """Raised when a visual-pipeline delivery is unsafe or incomplete."""


def _string(value, field: str, *, max_length: int = 256) -> str:
    if not isinstance(value, str) or not value.strip():
        raise CharacterAssetSchemaError(f"CharacterAsset {field} must be a non-empty string")
    clean = value.strip()
    if len(clean) > max_length:
        raise CharacterAssetSchemaError(f"CharacterAsset {field} is too long")
    return clean


def _number(value, field: str, minimum: float, maximum: float) -> float:
    if not isinstance(value, (int, float)) or isinstance(value, bool):
        raise CharacterAssetSchemaError(f"CharacterAsset {field} must be numeric")
    number = float(value)
    if not minimum <= number <= maximum:
        raise CharacterAssetSchemaError(
            f"CharacterAsset {field} must be in [{minimum}, {maximum}]"
        )
    return number


def _validate_glb_url(value) -> str:
    url = _string(value, "glb_url", max_length=2048)
    if ".." in url.replace("\\", "/").split("/"):
        raise CharacterAssetSchemaError("CharacterAsset glb_url cannot traverse directories")
    parsed = urlparse(url)
    if parsed.query or parsed.fragment or parsed.username or parsed.password:
        raise CharacterAssetSchemaError(
            "CharacterAsset glb_url must be immutable and cannot contain query, fragment, or credentials"
        )
    if parsed.scheme:
        if parsed.scheme != "https" or not parsed.netloc:
            raise CharacterAssetSchemaError("CharacterAsset glb_url only allows HTTPS")
    elif url.startswith("/"):
        if not url.startswith(_SAFE_ABSOLUTE_PREFIXES):
            raise CharacterAssetSchemaError("CharacterAsset absolute glb_url is outside allowed paths")
    elif not url.startswith(_SAFE_RELATIVE_PREFIXES):
        raise CharacterAssetSchemaError("CharacterAsset relative glb_url is outside allowed paths")
    if not parsed.path.lower().endswith(".glb"):
        raise CharacterAssetSchemaError("CharacterAsset glb_url must point to a .glb file")
    return url


def _validate_animations(value) -> dict[str, str]:
    if not isinstance(value, dict):
        raise CharacterAssetSchemaError("CharacterAsset runtime.animations must be an object")
    if len(value) > 32:
        raise CharacterAssetSchemaError("CharacterAsset runtime.animations has too many entries")
    animations = {}
    for semantic_name, clip_name in value.items():
        semantic = _string(semantic_name, "runtime.animations key", max_length=64)
        clip = _string(clip_name, f"runtime.animations.{semantic}", max_length=128)
        if "/" in clip or "\\" in clip or "://" in clip:
            raise CharacterAssetSchemaError(
                f"CharacterAsset runtime.animations.{semantic} must be a GLB clip name"
            )
        animations[semantic] = clip
    return animations


def validate_character_asset(asset) -> dict:
    """Validate and sanitize the only CharacterAsset fields allowed into browser DTOs."""
    if not isinstance(asset, dict):
        raise CharacterAssetSchemaError("CharacterAsset must be an object")
    if asset.get("schema_version") != SCHEMA_VERSION:
        raise CharacterAssetSchemaError(
            f"Unsupported CharacterAsset schema: {asset.get('schema_version')}"
        )
    character_id = _string(asset.get("character_id"), "character_id", max_length=128)
    revision = asset.get("revision")
    if not isinstance(revision, int) or isinstance(revision, bool) or revision < 1:
        raise CharacterAssetSchemaError("CharacterAsset revision must be an integer >= 1")
    content_hash = _string(asset.get("content_hash"), "content_hash", max_length=64)
    if not _SHA256.fullmatch(content_hash):
        raise CharacterAssetSchemaError("CharacterAsset content_hash must be a SHA-256 hex digest")

    runtime = asset.get("runtime")
    if not isinstance(runtime, dict):
        raise CharacterAssetSchemaError("CharacterAsset runtime must be an object")
    if runtime.get("forward_axis") != "+Z":
        raise CharacterAssetSchemaError("CharacterAsset runtime.forward_axis must be +Z")
    animations = _validate_animations(runtime.get("animations"))

    qa = asset.get("qa")
    if not isinstance(qa, dict) or qa.get("status") != "passed":
        raise CharacterAssetSchemaError("CharacterAsset qa.status must be passed")

    return {
        "schema_version": SCHEMA_VERSION,
        "character_id": character_id,
        "revision": revision,
        "glb_url": _validate_glb_url(asset.get("glb_url")),
        "content_hash": content_hash.lower(),
        "runtime": {
            "scale_meters": _number(runtime.get("scale_meters"), "runtime.scale_meters", 0.01, 3.0),
            "ground_offset": _number(runtime.get("ground_offset"), "runtime.ground_offset", -1.0, 1.0),
            "forward_axis": "+Z",
            "animations": animations,
        },
        "qa": {"status": "passed"},
    }

File: src/runtime/CharacterAsset.js
import { publicUrl } from "./WorldSpec.js";


export const CHARACTER_ASSET_SCHEMA = "character-asset.v1";

const SHA256_PATTERN = /^[a-f0-9]{64}$/i;
const SAFE_RELATIVE_PREFIXES = ["assets/", "models/"];
const SAFE_ABSOLUTE_PREFIXES = ["/assets/", "/models/", "/api/v0/assets/"];


function numberInRange(value, minimum, maximum) {
  return typeof value === "number" && Number.isFinite(value) && value >= minimum && value <= maximum
    ? value
    : null;
}


function safeGlbUrl(value) {
  if (typeof value !== "string" || !value.trim()) return null;
  const url = value.trim();
  if (url.length > 2048) return null;
  if (url.includes("..") || /[?#]/.test(url) || !url.toLowerCase().endsWith(".glb")) return null;
  if (url.startsWith("https://")) {
    try {
      const parsed = new URL(url);
      return parsed.username || parsed.password ? null : url;
    } catch {
      return null;
    }
  }
  if (url.startsWith("/")) {
    return SAFE_ABSOLUTE_PREFIXES.some((prefix) => url.startsWith(prefix)) ? url : null;
  }
  return SAFE_RELATIVE_PREFIXES.some((prefix) => url.startsWith(prefix)) ? url : null;
}


function safeAnimations(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) return null;
  const entries = Object.entries(value);
  if (entries.length > 32) return null;
  const animations = {};
  for (const [semanticName, clipName] of entries) {
    const semantic = semanticName.trim();
    if (!semantic || semantic.length > 64) return null;
    if (
      typeof clipName !== "string" ||
      !clipName.trim() ||
      clipName.trim().length > 128 ||
      /[\\/]|:\/\//.test(clipName)
    ) return null;
    animations[semantic] = clipName.trim();
  }
  return animations;
}


export function normalizeCharacterAsset(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) return null;
  if (value.schema_version !== CHARACTER_ASSET_SCHEMA) return null;
  if (
    typeof value.character_id !== "string" ||
    !value.character_id.trim() ||
    value.character_id.trim().length > 128
  ) return null;
  if (!Number.isInteger(value.revision) || value.revision < 1) return null;
  if (typeof value.content_hash !== "string" || !SHA256_PATTERN.test(value.content_hash)) return null;
  const glbUrl = safeGlbUrl(value.glb_url);
  if (!glbUrl) return null;

  const runtime = value.runtime;
  if (!runtime || typeof runtime !== "object" || Array.isArray(runtime)) return null;
  const scaleMeters = numberInRange(runtime.scale_meters, 0.01, 3);
  const groundOffset = numberInRange(runtime.ground_offset, -1, 1);
  if (scaleMeters === null || groundOffset === null || runtime.forward_axis !== "+Z") return null;
  const animations = safeAnimations(runtime.animations);
  if (!animations) return null;
  if (value.qa?.status !== "passed") return null;

  return Object.freeze({
    schema_version: CHARACTER_ASSET_SCHEMA,
    character_id: value.character_id.trim(),
    revision: value.revision,
    glb_url: glbUrl,
    content_hash: value.content_hash.toLowerCase(),
    runtime: Object.freeze({
      scale_meters: scaleMeters,
      ground_offset: groundOffset,
      forward_axis: "+Z",
      animations: Object.freeze(animations),
    }),
    qa: Object.freeze({ status: "passed" }),
  });
}


export function characterAssetKey(asset) {
  const normalized = normalizeCharacterAsset(asset);
  return normalized
    ? `${normalized.character_id}@${normalized.revision}:${normalized.content_hash}`
    : null;
}


export function characterAssetUrl(asset) {
  const normalized = normalizeCharacterAsset(asset);
  if (!normalized) return null;
  const url = normalized.glb_url;
  return url.startsWith("https://") || url.startsWith("/") ? url : publicUrl(url);
}
